File: Python implementation/models/member.py
"""
Member model for Library Management System

Represents a library member with personal information and borrowing history.
"""

from dataclasses import dataclass, field
from typing import List, Dict, Any


@dataclass
class Member:
    """
    Member data model.

    Attributes:
        member_id (str): Unique member identifier
        name (str): Member full name
        nrc (str): National Registration Card number
        phone (str): Contact phone number
        address (str): Residential address
        borrowed_books (List[Dict]): List of currently borrowed books with details
        fine_amount (float): Accumulated fines in MMK
    """
    member_id: str
    name: str
    nrc: str
    phone: str
    address: str
    borrowed_books: List[Dict[str, Any]] = field(default_factory=list)
    fine_amount: float = 0.0

    def __post_init__(self):
        """Validate data after initialization."""
        if self.fine_amount < 0:
            raise ValueError("Fine amount cannot be negative")

    def add_borrowed_book(self, book_id: str, book_title: str, borrow_date: str, due_date: str) -> None:
        """
        Add a book to the member's borrowed list.

        Args:
            book_id: Book identifier
            book_title: Book title
            borrow_date: Date when book was borrowed
            due_date: Date when book is due
        """
        borrow_record = {
            'book_id': book_id,
            'book_title': book_title,
            'borrow_date': borrow_date,
            'due_date': due_date
        }
        self.borrowed_books.append(borrow_record)

    def remove_borrowed_book(self, book_id: str) -> bool:
        """
        Remove a book from the member's borrowed list.

        Args:
            book_id: Book identifier to remove

        Returns:
            bool: True if book was found and removed, False otherwise
        """
        for i, book in enumerate(self.borrowed_books):
            if book['book_id'] == book_id:
                self.borrowed_books.pop(i)
                return True
        return False

    def has_borrowed_book(self, book_id: str) -> bool:
        """
        Check if member has borrowed a specific book.

        Args:
            book_id: Book identifier to check

        Returns:
            bool: True if member has borrowed the book
        """
        return any(book['book_id'] == book_id for book in self.borrowed_books)

    def get_borrowed_book_count(self) -> int:
        """
        Get the number of books currently borrowed by the member.

        Returns:
            int: Number of borrowed books
        """
        return len(self.borrowed_books)

    def add_fine(self, amount: float) -> None:
        """
        Add fine amount to member's account.

        Args:
            amount: Fine amount to add
        """
        if amount > 0:
            self.fine_amount += amount

    def clear_fine(self) -> None:
        """Clear all accumulated fines."""
        self.fine_amount = 0.0

    def __str__(self) -> str:
        """String representation of the member."""
        return f"{self.name} (ID: {self.member_id}) - Borrowed: {len(self.borrowed_books)} books"

    def __repr__(self) -> str:
        """Detailed string representation."""
        return f"Member(id='{self.member_id}', name='{self.name}', borrowed={len(self.borrowed_books)}, fine={self.fine_amount})"

File: Java implementation/Member.java
// Member.java - Member model

import java.util.*;

/**
 * Member class represents a library member.
 * 
 * Maintains member information and tracks borrowed books.
 */
public class Member {
    private String memberId;
    private String name;
    private String nrc;               // National Registration Card
    private String phone;
    private String address;
    private List<BorrowRecord> borrowedBooks;  // Currently borrowed
    private double fineAmount;
    
    /**
     * Inner class for tracking borrowed book records.
     */
    public static class BorrowRecord {
        public String bookId;
        public String bookTitle;
        public String borrowDate;
        public String dueDate;
        
        public BorrowRecord(String bookId, String bookTitle, String borrowDate, String dueDate) {
            this.bookId = bookId;
            this.bookTitle = bookTitle;
            this.borrowDate = borrowDate;
            this.dueDate = dueDate;
        }
    }
    
    /**
     * Constructor for Member.
     * 
     * @param memberId Unique member identifier
     * @param name Member full name
     * @param nrc National Registration Card number
     * @param phone Contact phone number
     * @param address Residential address
     */
    public Member(String memberId, String name, String nrc, String phone, String address) {
        if (nrc == null || nrc.isEmpty()) {
            throw new IllegalArgumentException("NRC cannot be empty");
        }
        
        this.memberId = memberId;
        this.name = name;
        this.nrc = nrc;
        this.phone = phone;
        this.address = address;
        this.borrowedBooks = new ArrayList<>();
        this.fineAmount = 0.0;
    }
    
    // Getters
    public String getMemberId() { return memberId; }
    public String getName() { return name; }
    public String getNrc() { return nrc; }
    public String getPhone() { return phone; }
    public String getAddress() { return address; }
    public List<BorrowRecord> getBorrowedBooks() { return borrowedBooks; }
    public double getFineAmount() { return fineAmount; }
    
    /**
     * Add a borrowed book record.
     * 
     * Time Complexity: O(1)
     * 
     * @param bookId Book identifier
     * @param bookTitle Book title
     * @param borrowDate Borrow date
     * @param dueDate Due date
     */
    public void addBorrowedBook(String bookId, String bookTitle, String borrowDate, String dueDate) {
        borrowedBooks.add(new BorrowRecord(bookId, bookTitle, borrowDate, dueDate));
    }
    
    /**
     * Remove a borrowed book record.
     * 
     * Time Complexity: O(k) where k = number of borrowed books
     * 
     * @param bookId Book identifier to remove
     * @return true if book was removed, false if not found
     */
    public boolean removeBorrowedBook(String bookId) {
        return borrowedBooks.removeIf(record -> record.bookId.equals(bookId));
    }
    
    /**
     * Check if member has borrowed a specific book.
     * 
     * Time Complexity: O(k)
     * 
     * @param bookId Book identifier to check
     * @return true if member has borrowed the book
     */
    public boolean hasBorrowedBook(String bookId) {
        return borrowedBooks.stream().anyMatch(record -> record.bookId.equals(bookId));
    }
    
    /**
     * Get number of books currently borrowed.
     * 
     * Time Complexity: O(1)
     * 
     * @return Number of borrowed books
     */
    public int getBorrowedBookCount() {
        return borrowedBooks.size();
    }
    
    /**
     * Add fine amount to member's account.
     * 
     * Time Complexity: O(1)
     * 
     * @param amount Fine amount to add
     */
    public void addFine(double amount) {
        if (amount > 0) {
            fineAmount += amount;
        }
    }
    
    /**
     * Clear all accumulated fines.
     * 
     * Time Complexity: O(1)
     */
    public void clearFine() {
        fineAmount = 0.0;
    }
    
    @Override
    public String toString() {
        return String.format("%s (ID: %s) - Borrowed: %d books, Fine: %.2f",
                           name, memberId, borrowedBooks.size(), fineAmount);
    }
}